x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/multiterms/MultiTermsAggregationBuilder.java

```diff
@@ -0,0 +1,430 @@
+    }
+
+    private List<MultiValuesSourceFieldConfig> terms = Collections.emptyList();
+
+    private BucketOrder order = BucketOrder.compound(BucketOrder.count(false));
+
+    boolean showTermDocCountError = false;
+
+    private Aggregator.SubAggCollectionMode collectMode = null;
+
+    private TermsAggregator.BucketCountThresholds bucketCountThresholds = new TermsAggregator.BucketCountThresholds(
+        DEFAULT_BUCKET_COUNT_THRESHOLDS
+    );
+
+    @FunctionalInterface
+    interface MultiTermValuesSupplier {
+        MultiTermsAggregator.TermValuesSource build(ValuesSourceConfig config);
+    }
+
+    static final ValuesSourceRegistry.RegistryKey<MultiTermValuesSupplier> REGISTRY_KEY = new ValuesSourceRegistry.RegistryKey<>(
+        NAME,
+        MultiTermValuesSupplier.class
+    );
+
+    public static void registerAggregators(ValuesSourceRegistry.Builder registry) {
+        registry.registerUsage(NAME);
+        registry.register(REGISTRY_KEY, List.of(CoreValuesSourceType.NUMERIC), MultiTermsAggregator::buildNumericTermValues, false);
+        registry.register(
+            REGISTRY_KEY,
+            List.of(CoreValuesSourceType.BOOLEAN, CoreValuesSourceType.DATE),
+            MultiTermsAggregator.LongTermValuesSource::new,
+            false
+        );
+        registry.register(REGISTRY_KEY, List.of(CoreValuesSourceType.KEYWORD), MultiTermsAggregator.StringTermValuesSource::new, false);
+        registry.register(REGISTRY_KEY, List.of(CoreValuesSourceType.IP), MultiTermsAggregator.IPTermValuesSource::new, false);
+    }
+
+    public MultiTermsAggregationBuilder(String name) {
+        super(name);
+    }
+
+    public MultiTermsAggregationBuilder(
+        MultiTermsAggregationBuilder clone,
+        AggregatorFactories.Builder factoriesBuilder,
+        Map<String, Object> metadata
+    ) {
+        super(clone, factoriesBuilder, metadata);
+        this.terms = new ArrayList<>(clone.terms);
+        this.order = clone.order;
+        this.collectMode = clone.collectMode;
+        this.showTermDocCountError = clone.showTermDocCountError;
+        this.bucketCountThresholds = new TermsAggregator.BucketCountThresholds(clone.bucketCountThresholds);
+    }
+
+    public MultiTermsAggregationBuilder(StreamInput in) throws IOException {
+        super(in);
+        terms = in.readList(MultiValuesSourceFieldConfig::new);
+        order = InternalOrder.Streams.readOrder(in);
+        collectMode = in.readOptionalWriteable(Aggregator.SubAggCollectionMode::readFromStream);
+        bucketCountThresholds = new TermsAggregator.BucketCountThresholds(in);
+        showTermDocCountError = in.readBoolean();
+    }
+
+    /**
+     * Sets the field to use for this aggregation.
+     */
+    public MultiTermsAggregationBuilder terms(List<MultiValuesSourceFieldConfig> terms) {
+        if (terms == null) {
+            throw new IllegalArgumentException("[terms] must not be null: [" + name + "]");
+        }
+        this.terms = terms;
```

Comment:
I am checking the size in doBuild, but I will move it here, no reason to delay the check for so long. 

If a user specifies the same field twice we will add it twice, it only makes sense if the field contains multiple values (we count all possible pairs) and I am not sure what the practical value for using it, but since it looks like a potentially useful functionality and doesn't cost us anything, I decided not to prohibit it. I am not trying to optimize for it ether.

Reply:
Ooh, didn't think about multi-valued fields.  Yeah seems a really odd use-case but no need to disallow it either.  :+1: